accord-core/src/main/java/accord/impl/AbstractFetchCoordinator.java

```diff
@@ -0,0 +1,288 @@
+
+    static class FetchResult extends AsyncResults.SettableResult<Ranges> implements DataStore.FetchResult
+    {
+        final AbstractFetchCoordinator coordinator;
+
+        FetchResult(AbstractFetchCoordinator coordinator)
+        {
+            this.coordinator = coordinator;
+        }
+
+        @Override
+        public void abort(Ranges abort)
+        {
+            coordinator.abort(abort);
+        }
+    }
+
+    static class Key
+    {
+        final Node.Id id;
+        final Ranges ranges;
+
+        Key(Node.Id id, Ranges ranges)
+        {
+            this.id = id;
+            this.ranges = ranges;
+        }
+
+        @Override
+        public int hashCode()
+                         {
+                            return id.hashCode() + ranges.hashCode();
+                                                                     }
+
+        @Override
+        public boolean equals(Object obj)
+        {
+            if (this == obj) return true;
+            if (!(obj instanceof Key)) return false;
+            Key that = (Key) obj;
+            return id.equals(that.id) && ranges.equals(that.ranges);
+        }
+    }
+
+    final DataStore.FetchRanges fetchRanges;
+    final CommandStore commandStore;
+    final Map<Key, DataStore.StartingRangeFetch> inflight = new HashMap<>();
+    final FetchResult result = new FetchResult(this);
+    final List<AsyncResult<Void>> persisting = new ArrayList<>();
+
+    protected AbstractFetchCoordinator(Node node, Ranges ranges, SyncPoint syncPoint, DataStore.FetchRanges fetchRanges, CommandStore commandStore)
+    {
+        super(node, ranges, syncPoint, fetchRanges);
+        this.fetchRanges = fetchRanges;
+        this.commandStore = commandStore;
+    }
+
+    protected abstract PartialTxn rangeReadTxn(Ranges ranges);
+
+    protected abstract void onReadOk(Node.Id from, CommandStore commandStore, Data data, Ranges ranges);
+
+    @Override
+    public void contact(Node.Id to, Ranges ranges)
+    {
+        Key key = new Key(to, ranges);
+        inflight.put(key, starting(to, ranges));
+        Ranges ownedRanges = ownedRangesForNode(to);
+        Invariants.checkArgument(ownedRanges.containsAll(ranges));
+        PartialDeps partialDeps = syncPoint.waitFor.slice(ownedRanges, ranges);
+        node.send(to, new FetchRequest(syncPoint.sourceEpoch(), syncPoint.syncId, ranges, partialDeps, rangeReadTxn(ranges)), new Callback<ReadData.ReadReply>()
+        {
+            @Override
+            public void onSuccess(Node.Id from, ReadData.ReadReply reply)
+            {
+                if (!reply.isOk())
+                {
+                    fail(to, new RuntimeException(reply.toString()));
+                    inflight.remove(key).cancel();
+                    switch ((ReadData.ReadNack) reply)
+                    {
+                        default: throw new AssertionError("Unhandled enum");
+                        case Invalid:
+                        case Redundant:
```

Comment:
trying to wrap my head around this... doesn't this just mean concurrent txn (original coordinator + recovery coordinator)?  So this is just saying that "we are reusing message reply, and these states are not valid for us" right?